csharp/src/Ice/Communicator-LocatorManager.cs

```diff
@@ -12,437 +13,366 @@ namespace ZeroC.Ice
-
-                if (_response)
-                {
-                    callback.Response(LocatorInfo, _proxy);
                 }
                 else
                 {
-                    Debug.Assert(_exception != null);
-                    callback.Exception(_exception);
+                    Communicator communicator = reference.Communicator;
+                    var s = new System.Text.StringBuilder();
+                    s.Append("no endpoints configured for ");
+                    if (reference.AdapterId.Length > 0)
+                    {
+                        s.Append("adapter\n");
+                        s.Append("adapter = " + reference.AdapterId);
+                    }
+                    else
+                    {
+                        s.Append("well-known object\n");
+                        s.Append("well-known proxy = " + reference.ToString());
+                    }
+                    communicator.Logger.Trace(communicator.TraceLevels.LocationCat, s.ToString());
                 }
             }
+            return (endpoints ?? Array.Empty<Endpoint>(), cached);
+        }
 
-            internal void Response(IObjectPrx? proxy)
+        private async Task<IReadOnlyList<Endpoint>?> GetAdapterEndpointsAsync(Reference reference)
+        {
+            if (reference.Communicator.TraceLevels.Location > 0)
             {
-                lock (_mutex)
-                {
-                    LocatorInfo.FinishRequest(Ref, _wellKnownRefs, proxy, false);
-                    _response = true;
-                    _proxy = proxy;
-                    Monitor.PulseAll(_mutex);
-                }
-                foreach (RequestCallback callback in _callbacks)
-                {
-                    callback.Response(LocatorInfo, proxy);
-                }
+                reference.Communicator.Logger.Trace(reference.Communicator.TraceLevels.LocationCat,
+                    $"searching for adapter by id\nadapter = {reference.AdapterId}");
             }
 
-            internal void Exception(Exception ex)
+            Task<IReadOnlyList<Endpoint>?>? task;
+            lock (_mutex)
             {
-                lock (_mutex)
-                {
-                    LocatorInfo.FinishRequest(Ref, _wellKnownRefs, null, ex is RemoteException);
-                    _exception = ex;
-                    Monitor.PulseAll(_mutex);
-                }
-                foreach (RequestCallback callback in _callbacks)
+                if (!_adapterRequests.TryGetValue(reference.AdapterId, out task))
                 {
-                    callback.Exception(ex);
+                    // If there's no locator request in progress for this adapter, we invoke one and cache it to prevent
+                    // making too many requests on the locator. It's removed once the locator response is received.
+                    task = PerformGetAdapterEndpointsAsync(reference);
+                    if (!task.IsCompleted)
```

Comment:
Hmm, what if the sending of the underlying FindAdapterById two-way request is performed synchronously and the response is received before the await on the FindAdapterById response task completion source? The await for the response would complete synchronously as well in this case. 

The await for FindAdapterById would therefore complete synchronously which would cause PerformGetAdapterEndpointsAsync to return a completed task. What am I missing?

Reply:
You're right. It would not complete synchronously but it could complete by the time we reach this if (it's unlikely though), and adding the completed task to the list would create a leak.

Reply:
Actually, since:
 - `GetAdapterEndpointsAsync` locks _mutex the entire time
 - the finally of `PerformGetAdapterEndpointsAsync` task lock the same _mutex

I don't see how it can complete until the mutex is released.

Reply:
isn't it because the C# mutex is recursive?

Reply:
Yes, the mutex acquisition from the finally won't block when the Ice calls completes synchronously since called from the thread that locks it already when calling PerformGetAdapterEndpointsAsync... another way to write it would be to make the more explicit from the caller by removing the finally block from PerformGetAdapterEndpointsAsync and instead do the remove from another GetAadapterEndpointsAsync method as bellow:
```

Task<IReadOnlyList<Endpoint>>? task;
lock (_mutex)
{
    if (!_adapterRequests.TryGetValue(reference.AdapterId, out task))
    {
        // If there's no locator request in progress for this adapter, we invoke one and cache it to prevent
        // making too many requests on the locator. It's removed once the locator response is received.
        task = PerformGetAdapterEndpointsAsync(reference);
        if (!task.IsCompleted)
        {
            _adapterRequests.Add(reference.AdapterId, task);
            _ = RemoveAdapterRequestAfterCompletionAsync(reference.AdapterId, task);
        }
    }
}
return task!;

async Task RemoveAfterCompletionAsync(string adapterId, Task<IReadOnlyList<Endpoint>> task)
{
    try
    {
        await task;
    }
    catch (System.Exception)
    {
        lock (_mutex)
        {
            _adapterRequests.Remove(adapterId);
        }
        throw;
    }
}
```

More core for no real benefits however... I think I prefer adding an additional comment, hopefully that will be clearer.

Reply:
I think adding a comment will be enough, is the alternate correct why remove the request only when there is an exception maybe the catch was supposed to be a finally?

Reply:
The catch is indeed supposed to be a finally 😄 but yes, a comment should be good enough.

Reply:
For me, this cannot complete _successfully_ synchronously:
```
IObjectPrx? proxy = await Locator.FindAdapterByIdAsync(reference.AdapterId).ConfigureAwait(false);
```
i.e. if this call succeeds, the continuation always executes in a separate thread - not the thread that called `PerformGetAdapterEndpointsAsync` and that has _mutex locked.

So this is all about the situation where this call actually completes with a failure synchronously?

Reply:
Hmm what if we rewrite it as follow:
```
var task = Locator.FindAdapterByIdAsync(reference.AdapterId);
await task.ConfigureAwait();
```
Isn't it equivalent? I don't see why the task couldn't complete between these two lines and therefore cause the await to return immediately. What am I missing?

Reply:
Yes, it's equivalent, and this synchronous completion seems indeed possible, but unlikely.
Assuming `findAdapterById` is sent synchronously by `FindAdapterByIdAsync`, its response will be necessarily received by a separate thread. And all this would need to happen between these two lines.

That's why we use Tasks for the AMI API: synchronous completely if (very) unlikely, but not totally impossible.

Reply:
It's more likely if you consider collocated locator implementations where no networking is involved (but the response is still sent by a separate thread) such as IceDiscovery or IceLocatorDiscovery.

Regarding the AMI mapping, Note that InvokeAsync uses ValueTask right now... (I don't know why). Also note that oneway invocations are likely to complete synchronously and ValueTask could be cheaper in this case. There would be no Task creation at all involved on the invocation code path... Task is probably safer however and probably not that expensive for an Ice invocation (but this doesn't mean we shouldn't be careful and try to optimize the impl. of the invocation under the hood to use ValueTask... such as for `Reference.GetRequestHandlerAsync`).